dot/state/grandpa.go

```diff
@@ -26,42 +37,477 @@ var (
+	genesisAuthorities []types.GrandpaVoter) (*GrandpaState, error) {
 	grandpaDB := chaindb.NewTable(db, grandpaPrefix)
 	s := &GrandpaState{
-		db: grandpaDB,
+		db:         grandpaDB,
+		blockState: bs,
 	}
 
 	if err := s.setCurrentSetID(genesisSetID); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set current set id: %w", err)
 	}
 
 	if err := s.SetLatestRound(0); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set latest round: %w", err)
 	}
 
 	if err := s.setAuthorities(genesisSetID, genesisAuthorities); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set authorities: %w", err)
 	}
 
-	if err := s.setSetIDChangeAtBlock(genesisSetID, 0); err != nil {
-		return nil, err
+	if err := s.setChangeSetIDAtBlock(genesisSetID, 0); err != nil {
+		return nil, fmt.Errorf("cannot set change set id at block 0: %w", err)
 	}
 
 	return s, nil
 }
 
 // NewGrandpaState returns a new GrandpaState
-func NewGrandpaState(db chaindb.Database) (*GrandpaState, error) {
+func NewGrandpaState(db chaindb.Database, bs *BlockState) (*GrandpaState, error) {
 	return &GrandpaState{
-		db: chaindb.NewTable(db, grandpaPrefix),
+		db:         chaindb.NewTable(db, grandpaPrefix),
+		blockState: bs,
 	}, nil
 }
 
+// HandleGRANDPADigest receives a decoded GRANDPA digest and calls the right function to handles the digest
+func (s *GrandpaState) HandleGRANDPADigest(header *types.Header, digest scale.VaryingDataType) error {
+	switch val := digest.Value().(type) {
+	case types.GrandpaScheduledChange:
+		return s.addScheduledChange(header, val)
+	case types.GrandpaForcedChange:
+		return s.addForcedChange(header, val)
+	case types.GrandpaOnDisabled:
+		return nil
+	case types.GrandpaPause:
+		logger.Warn("GRANDPA Pause consensus message not imeplemented yet")
+		return nil
+	case types.GrandpaResume:
+		logger.Warn("GRANDPA Resume consensus message not imeplemented yet")
+		return nil
+	default:
+		return fmt.Errorf("not supported digest")
+	}
+}
+
+func (s *GrandpaState) addForcedChange(header *types.Header, fc types.GrandpaForcedChange) error {
+	headerHash := header.Hash()
+
+	for _, change := range s.forcedChanges {
+		changeBlockHash := change.announcingHeader.Hash()
+
+		if changeBlockHash.Equal(headerHash) {
+			return errDuplicateHashes
+		}
+
+		isDescendant, err := s.blockState.IsDescendantOf(changeBlockHash, headerHash)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			return errAlreadyHasForcedChanges
+		}
+	}
+
+	auths, err := types.GrandpaAuthoritiesRawToAuthorities(fc.Auths)
+	if err != nil {
+		return fmt.Errorf("cannot parser GRANDPA authorities to raw authorities: %w", err)
+	}
+
+	pendingChange := &pendingChange{
+		bestFinalizedNumber: fc.BestFinalizedBlock,
+		nextAuthorities:     auths,
+		announcingHeader:    header,
+		delay:               fc.Delay,
+	}
+
+	s.forcedChanges = append(s.forcedChanges, pendingChange)
+	sort.Sort(s.forcedChanges)
```

Comment:
If the forcedChanges are already sorted before this append, you could do a more efficient sorted insertion, see https://stackoverflow.com/a/55460931